[MRG + 1] Ensure dict_learning_online has at least one batch

betatim opened the pull request:
If there are less samples than the batch_size set the number of batches to one. This came up during #4694. Is this a good fix? Are there more tests which special case this?

arjoly commented:
I would try to patch / use the [gen_batches](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/__init__.py#L368) or [gen_even_slices](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/__init__.py#L393) functions.

betatim commented:
Thanks, `gen_batches` is nice.

Should we keep the warning? It seems if you delegate the chunking up `gen_batches` that is the one that should do the warning (if any).

amueller commented:
I think if ``batch_size > n_samples`` then ``gen_batches`` just generates a single batch, right?
Which is what I think it should do. I don't think we need to add a warning.

betatim commented:
Done, warning removed.

I don't understand why one of the CIs [failed](https://travis-ci.org/scikit-learn/scikit-learn/jobs/62182759#L1537). It is somewhere in `linear_model`, from a quick glance it does not use any of the code I touched and it is just on one of the builds. Any ideas?

amueller commented:
Unrelated heisenfailure. I opened an issue for it here: #4719.
Looks good to me. Can you squash?

betatim committed:
Ensure dict_learning_online always has at least one batch

Now uses gen_batches to create batches which works even
when n_samples < batch_size. Removes need for special casing
MiniBatchDictLearning and MiniBatchSparsePCA in the tests.

arjoly commented:
LGTM

amueller commented:
Thanks @betatim :)

betatim commented:
Avec plaisir
